src/MapLibreBasedGeocodingControl.ts

```diff
@@ -87,26 +95,156 @@ export type MapLibreBaseControlOptions = Omit<ControlOptions, "apiKey"> & {
 
 export type Props<T> = T extends SvelteComponent<infer P> ? P : never;
 
+type RemoveString<T> = T extends string ? never : T;
+
+type Event$1 = RemoveString<Parameters<Evented["fire"]>[0]>;
+
 type EventedConstructor = new (
   ...args: ConstructorParameters<typeof Evented>
 ) => Evented;
 
-export function crateBaseClass(
+export function crateClasses<OPTS extends MapLibreBaseControlOptions>(
   Evented: EventedConstructor,
   maplibreGl: MapLibreGL,
   getExtraProps?: (
     map: Map,
     div: HTMLElement,
   ) => Partial<Props<GeocodingControlComponent>>,
 ) {
-  return class MapLibreBasedGeocodingControl<
-    T extends MapLibreBaseControlOptions,
-  > extends Evented {
+  type EventTypes = {
+    select: SelectEvent;
+    featureslisted: FeaturesListedEvent;
+    featuresmarked: FeaturesMarkedEvent;
+    optionsvisibilitychange: OptionsVisibilityChangeEvent;
+    pick: PickEvent;
+    querychange: QueryChangeEvent;
+    response: ResponseEvent;
+    reversetoggle: ReverseToggleEvent;
+  };
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const _keysMatch: EnsureMatchingKeys<EventTypes, DispatcherType> = true;
```

Comment:
since we don't use it - do we need it?

Reply:
More context: https://chatgpt.com/share/673b3271-2ab8-8007-b4c2-15595f32b0ec

Reply:
oh, now I see! Let me do a quick test

Reply:
So as far I understand, we don't want to simply use `EvenTypes = DispatcherType` to make EvenTypes fields type more precise.

I think this does what we want (and event makes sure that field type is acceptable for OriginalType):
```
  type RedefineType<OriginalType, UpdatedType extends { [K in keyof OriginalType]: OriginalType[K] }> = UpdatedType;

  type EventTypes = RedefineType<DispatcherType, {
    select: SelectEvent;
    featureslisted: FeaturesListedEvent;
    featuresmarked: FeaturesMarkedEvent;
    optionsvisibilitychange: OptionsVisibilityChangeEvent;
    pick: PickEvent;
    querychange: QueryChangeEvent;
    response: ResponseEvent;
    reversetoggle: ReverseToggleEvent;
  }>;
  ```
That way we get an error if any of key is missing in declaration of EventTypes.
But if there is any edgecase I don't see, then `EnsureMatchingKeys` is fine 👍

Reply:
I like it. I've extended the type to also prevent adding extra properties:

```ts
type RedefineType<
  OriginalType,
  UpdatedType extends { [K in keyof OriginalType]: OriginalType[K] } & {
    [K in Exclude<keyof UpdatedType, keyof OriginalType>]: never;
  },
> = UpdatedType;
```